AvalonStudio/AvalonStudio.Extensibility/Utils/FuzzyMatch.cs

```diff
@@ -0,0 +1,448 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace AvalonStudio.Extensibility.Utils
+{
+
+    public class FuzzyMatch
+    {
+        static public (List<int> specials, int lastSegmentSpecialsIndex) FindSpecialCharacters(string str)
+        {
+            const string special_chars = ".-_";
+            int i;
+            char c;
+            int lastSegmentSpecialsIndex = 0;
+
+            List<int> specials = new List<int>();
+            specials.Add(0);    // The beginning of a string is always special!
+
+            bool lastWasLower = false;
+
+            for (i = 0; i < str.Length; i++)
+            {
+                c = str[i];
+
+                if (c == '/' || c == '\\')
+                {
+                    specials.Add(i++);
+                    specials.Add(i);
+                    lastSegmentSpecialsIndex = specials.Count - 1;
+                    lastWasLower = false;
+                }
+                else if (special_chars.IndexOf(c) >= 0)
+                {
+                    specials.Add(i++);
+                    specials.Add(i);
+                    lastWasLower = false;
+                }
+                else if (Char.ToUpper(c) == c)
+                {
+                    if (lastWasLower)
+                    {
+                        specials.Add(i);
+                    }
+
+                    lastWasLower = false;
+                }
+                else
+                {
+                    lastWasLower = true;
+                }
+            }
+
+            return (specials, lastSegmentSpecialsIndex);
+        }
+
+        public enum States
+        {
+            SPECIALS_MATCH,
+            ANY_MATCH,
+        };
+
+        public class Match
+        {
+            public Match(int index, bool isSpecial)
+            {
+                this.index = index;
+                this.isSpecial = isSpecial;
+            }
+
+            public int index;
+            public bool isSpecial;
+        };
+
+        public static List<Match> generateMatchList(string query, string str, List<int> specials, int startingSpecial)
```

Comment:
method name should be caps case. 

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members